### PR TITLE
Allow spam-checker modules to be async

### PR DESCRIPTION
**Pull request by Yoric**
The methods of spam checker plugins can now be `async`. This also consistently uses the `maybe_awaitable` helper in some more places (generally where we accept helpers that may or may not return `Awaitables`).

This will let us write spam checkers that need to hit a database or an external service before responding.

### CODE REVIEW

File: synapse/federation/federation_base.py [hunk 1]
```diff
@@ -78,7 +79,7 @@ def _check_sigs_and_hashes(
 
         ctx = current_context()
 
-        def callback(_, pdu: EventBase):
+        async def callback(_, pdu: EventBase):
```
Comment (clokep) [inline synapse/federation/federation_base.py:82]: I think this is the issue. A `Deferred`, doesn't know what to do with awaitables, I think we can either make this an `inlineCallbacks` call or restructure this code deeper. Nothing obvious is coming to mind at the moment though.

**Comment by richvdh** [comment]
@Yoric as a general rule it's best to keep the first line of a git commit comment to 50 characters or less; they get truncated above a certain limit anyway. https://chris.beams.io/posts/git-commit/#limit-50 has some more info here. 

I've fixed the subject of this PR.

**Comment by Yoric** [comment]
> @Yoric as a general rule it's best to keep the first line of a git commit comment to 50 characters or less; they get truncated above a certain limit anyway. https://chris.beams.io/posts/git-commit/#limit-50 has some more info here.
> 
> I've fixed the subject of this PR.

Oops, got to unlearn a few things from mercurial, thanks :)

File: synapse/federation/federation_base.py (hunk 1 above)
Comment (Yoric) [reply synapse/federation/federation_base.py:82]: I *think* I fixed that, but I still get the exact same error messages, so... I guess I didn't actually fix it.

**Comment by clokep** [comment]
A couple of other random thoughts on this:

* Would we want to allow any of the spam check methods to be async?
* I'm not sure we should be using `maybe_awaitable`, it seems rather unused in our code.

**Comment by Yoric** [comment]
> A couple of other random thoughts on this:
> 
>     * Would we want to allow any of the spam check methods to be async?

If it's not too much work, definitely.

>     * I'm not sure we should be using `maybe_awaitable`, it seems rather unused in our code.

**Comment by clokep** [comment]
I realized this could be simplified a bit since we can hide all the "is the spam check plugin" async inside of `synapse.events.spamcheck.SpamChecker` class and make all the methods in that async. (Since that class itself isn't pluggable, just the under-lying checkers...yes the naming is confusing.)

File: synapse/util/async_helpers.py [hunk 2]
```diff
@@ -525,28 +527,11 @@ def failure_cb(val):
     return new_d
 
 
-@attr.s(slots=True, frozen=True)
-class DoneAwaitable:
-    """Simple awaitable that returns the provided value.
+async def maybe_awaitable(value: Union[Awaitable[R], R]) -> R:
```
Comment (clokep) [inline synapse/util/async_helpers.py:530]: This seems much simpler, plus using the `isawaitable` call seems more inline with recommendations.

File: synapse/federation/federation_base.py [hunk 3]
```diff
@@ -78,6 +78,7 @@ def _check_sigs_and_hashes(
 
         ctx = current_context()
 
+        @defer.inlineCallbacks
```
Comment (clokep) [inline synapse/federation/federation_base.py:81]: I mentioned this in the commit, but it is sad to add an inlineCallbacks back in, but without this we need to convert a huge stack to async/await. I'd rather do that as a follow-up.

File: synapse/util/async_helpers.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/util/async_helpers.py:530]: It will mean that we add `maybe_awaitable` to the call stack, which is a bit sad.

Comment (clokep) [reply synapse/util/async_helpers.py:530]: Ah, I see. I can revert this change if you'd prefer?

File: synapse/util/async_helpers.py [hunk 4]
```diff
@@ -542,11 +544,11 @@ def __next__(self):
         raise StopIteration(self.value)
 
 
-def maybe_awaitable(value):
+def maybe_awaitable(value: Union[Awaitable[R], R]) -> Awaitable[R]:
     """Convert a value to an awaitable if not already an awaitable.
     """
-
-    if hasattr(value, "__await__"):
+    if inspect.isawaitable(value):
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/util/async_helpers.py:550]: Yup, fair

**Pull request merged**